src/stdlib_stringlist.f90

```diff
@@ -0,0 +1,726 @@
+! stdlib_stringlist.f90 --
+!     Module for storing and manipulating list of strings
+!     The strings may have arbitrary lengths, not necessarily the same
+!
+!     insert AT:      Inserts an element BEFORE the element present currently at the asked index
+!                       for forward indexes, otherwise
+!                     Inserts an element AFTER the element present currently at the asked index
+!                       for backward indexes
+!                     In other words, after insertion the element will be present at the asked index
+!                       for both forward and backward indexes                    
+!     insert BEFORE:  Inserts an element BEFORE the element present currently at the asked index
+!     insert AFTER:   Inserts an element AFTER the element present currently at the asked index
+!
+!     Note the distinction between AT and BEFORE in the module. Care has been taken to keep it consistent
+!     throughout the PR
+!
+module stdlib_stringlist
+    use stdlib_string_type, only: string_type, operator(/=) !, move
+    use stdlib_math, only: clip
+    ! use stdlib_optval, only: optval
+    implicit none
+    private
+
+    public :: stringlist_type, operator(//), operator(==), operator(/=)
+    public :: list_head, list_tail, fidx, bidx, stringlist_index_type
+
+    type stringlist_index_type
+        private
+        logical :: forward
+        integer :: offset
+
+    end type stringlist_index_type
+
+    type(stringlist_index_type), parameter :: list_head     = stringlist_index_type( .true. , 1 )   ! fidx(1)
+    type(stringlist_index_type), parameter :: list_tail     = stringlist_index_type( .false., 1 )   ! bidx(1)
+
+    !> Version: experimental
+    !> 
+    !> Returns an instance of type 'stringlist_index_type' representing forward index
+    !> [Specifications](../page/specs/stdlib_stringlist.html#fidx)
+    interface fidx
+        module procedure forward_index
+    end interface
+
+    !> Version: experimental
+    !> 
+    !> Returns an instance of type 'stringlist_index_type' representing backward index
+    !> [Specifications](../page/specs/stdlib_stringlist.html#bidx)
+    interface bidx
+        module procedure backward_index
+    end interface
+
+    type stringlist_type
```

Comment:
I don't see any "constructors" for this type (i.e. overload `stringlist_type()`). Must one always build up a list imperatively and mutably? I tend to prefer a declarative, immutable style (i.e. once I construct an object, I don't mutate it).

Reply:
Yeah that's right I forgot to add that.
A constructor which can take an array as input and which can take no argument is what I am planning to add. I will make the changes.

Reply:
From my understanding of `stringlist_type` there is no need for initialization, you should be able to start right away with the default constructor and can just append using `operator(//)` to the instance.

The question is whether we want to additionally support `stringlist_type(["1", "2", "3"])` instead of requiring the usage of `stringlist_type() // ["1", "2", "3"]`?

Reply:
At the time I wrote that comment, I thought the default structure constructor was private, since the components are private, but since all the components either 1) have default initialization, 2) are allocatable, they are optional in the default structure constructor, and so the default structure constructor with no arguments is still public. The nuances of some of this are hard. At any rate, I think it would still be nice to support `stringlist_type(["1", "2", "3"])`.

Reply:
added the constructor which will take carray (char array) as well as sarray (string array) as input.